### PR TITLE
ARROW-8147: [C++] Add google-cloud-cpp to ThirdpartyToolchain

### PR DESCRIPTION
**Pull request by josiahyan**
I'm hoping to work on parts of the GCS filesystem implementation, and this is a prerequisite.

### CODE REVIEW

**Comment by josiahyan** [comment]
Running the cmake autoformatter heavily modifies the touched cmake files; should I be doing that?

**Comment by xhochy** [comment]
> Running the cmake autoformatter heavily modifies the touched cmake files; should I be doing that?

Maybe you have a different version of that installed and that formats differently? You can also summon the autoformatter with `@github-actions<space>autotune` on the PR.

**Comment by josiahyan** [comment]
I'll take a look! Didn't know I could do that.

**Comment by josiahyan** [comment]
@github-actions autotune

**Comment by josiahyan** [comment]
@xhochy I can't seem summon the bot. But I read what the action did, and saw that it required a very specific version of the cmake autoformatter, as you suggested. It works now! Thanks for your help!

**Comment by xhochy** [comment]
> @xhochy I can't seem to summon the bot. But I read what the action did, and saw that it required a very specific version of the cmake autoformatter, as you suggested. It works now! Thanks for your help!

The bot did run but didn't make any changes to your CMake files: https://github.com/apache/arrow/runs/1450988053?check_suite_focus=true#step:10:11

**Comment by josiahyan** [comment]
Oh, I know why! The CI queue was full (my test runs were taking hours to go through). By the time the runner got to my PR (Wed, 25 Nov 2020 02:15:47 GMT), my final commit had gone in.

File: ci/docker/conda-cpp.dockerfile [hunk 1]
```diff
@@ -73,6 +73,7 @@ ENV ARROW_BUILD_TESTS=ON \
     ARROW_PARQUET=ON \
     ARROW_PLASMA=ON \
     ARROW_S3=ON \
+    ARROW_GCS=ON \
```
Comment (pitrou) [inline ci/docker/conda-cpp.dockerfile:76]: Can we keep alphebetical order? Same for other changes below.

Comment (josiahyan) [reply ci/docker/conda-cpp.dockerfile]: Thanks for reviewing! I've reordered them.

**Comment by josiahyan** [comment]
@github-actions rebase

File: cpp/thirdparty/versions.txt [hunk 2]
```diff
@@ -32,14 +32,17 @@ ARROW_BOOST_BUILD_VERSION=1.71.0
...
 ARROW_GBENCHMARK_BUILD_VERSION=v1.5.2
+ARROW_GCPSDK_BUILD_VERSION=v1.20.0
 ARROW_GFLAGS_BUILD_VERSION=v2.2.2
 ARROW_GLOG_BUILD_VERSION=v0.4.0
 ARROW_GRPC_BUILD_VERSION=v1.29.1
 ARROW_GTEST_BUILD_VERSION=1.8.1
 ARROW_JEMALLOC_BUILD_VERSION=5.2.1
 ARROW_LZ4_BUILD_VERSION=v1.9.2
 ARROW_MIMALLOC_BUILD_VERSION=v1.6.4
+ARROW_NLOHMANN_JSON_BUILD_VERSION=v3.9.0
```
Comment (nealrichardson) [inline cpp/thirdparty/versions.txt:45]: Should URLs for these new dependencies also be added to `DEPENDENCIES` below? cf. ARROW-6407 ARROW-6333

Comment (josiahyan) [reply cpp/thirdparty/versions.txt:45]: Whoops, thanks! I missed that. I've added the URLs for the dependencies I've inserted into `versions.txt`.

However, I've left out the backup URLs that use the ursa-labs archive in `ThirdPartyDependencies.cmake`, which I understand will be derived from the `versions.txt` URL eventually (if this gets merged). Let me know if its appropriate to link add these URLs.

Comment (nealrichardson) [reply cpp/thirdparty/versions.txt:45]: No, it's fine, don't add URLs that don't exist

**Comment by nealrichardson** [comment]
Hi @josiahyan, apologies for the delay here. Could you please rebase this? Hopefully we can get this reviewed and merged soon after.

**Comment by josiahyan** [comment]
Whoops no problem; I’ll rebase it tonight or over the weekend. Thanks for
picking this up again!

**Comment by josiahyan** [comment]
Sorry about the delay! I've rebased the commits and manually merged in the changes made in master.

File: ci/conda_env_unix.yml [hunk 3]
```diff
@@ -19,6 +19,8 @@
 
 autoconf
 ccache
+google-cloud-cpp=1.20.0
```
Comment (xhochy) [inline ci/conda_env_unix.yml:22]: What is the reason for the exact pin here?

Comment (josiahyan) [reply ci/conda_env_unix.yml:22]: The currently available version in conda (1.24) is incompatible with the way I initially packaged it. I believe I saw in the CI run that `storage_client` was missing or something, as the google-cloud-cpp package has [started namespacing its exports](https://github.com/googleapis/google-cloud-cpp/blob/5bf147226e9333c5106d5fc8b8bba62657eeab97/google/cloud/storage/legacy.cmake.in#L28) (finally!).

Given the hard link against abseil libraries (as gRPC does too), I decided that the version was safer pinned.

I'm hoping to upgrade the dependency when there is code using it (which would be the GCS implementation of the Arrow file interfaces).

Comment (kou) [reply ci/conda_env_unix.yml:22]: I determined required abseil libraries manually. CI error helped me.

**Comment by salqadri** [comment]
What's the status of this PR?

**Comment by josiahyan** [comment]
> What's the status of this PR?

Sorry, but I lost track of following up on this PR! I believed it was ready to merge in Feb, but assumed people had lost interest in pursuing (native) GCS support, or that this change was not quite worth making for now. Is there still a need for this? If so, I can attempt to rebase it again.

**Comment by nealrichardson** [comment]
> > What's the status of this PR?
> 
> Sorry, but I lost track of following up on this PR! I believed it was ready to merge in Feb, but assumed people had lost interest in pursuing (native) GCS support, or that this change was not quite worth making for now. Is there still a need for this? If so, I can attempt to rebase it again.

I think there's interest (I'm interested, at least). I'd like to see someone more skilled in cmake than I to review and approve: @kou @pitrou @xhochy ?

**Comment by kou** [comment]
Could you rebase on the master? Then I'll review this.

**Comment by josiahyan** [comment]
Will rebase!

**Comment by emkornfield** [comment]
@josiahyan do you think you will have time to rebase?  Otherwise I can see how difficult this might be.

**Comment by josiahyan** [comment]
> @josiahyan do you think you will have time to rebase? Otherwise I can see how difficult this might be.

Whoops sorry, I lost track of this. I’ll be trying to rebase this - I don’t see any major changes (so far) that should result in conflicts.

File: ci/conda_env_unix.txt [hunk 4]
```diff
@@ -19,5 +19,7 @@
 
 autoconf
 ccache
+google-cloud-cpp=1.20.0
+nlohmann_json>=3.4.0
```
Comment (emkornfield) [inline ci/conda_env_unix.txt:23]: what is the new JSON dependency for?

Comment (emkornfield) [reply ci/conda_env_unix.txt:23]: I guess google's SDK must rely on it?

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 5]
```diff
@@ -2940,6 +3013,190 @@ if(ARROW_S3)
...
+  if(CARES_VENDORED)
+    list(APPEND GCPSDK_PREFIX_PATH_LIST ${CARES_PREFIX})
+  endif()
+
+  set(GCPSDK_PREFIX_PATH_LIST_SEP_CHAR "|")
+  list(JOIN GCPSDK_PREFIX_PATH_LIST ${GCPSDK_PREFIX_PATH_LIST_SEP_CHAR}
+       GCPSDK_PREFIX_PATH)
+
+  set(GCPSDK_CXX_FLAGS "${EP_CXX_FLAGS}")
+  # workaround for crc32c not declaring its header interface in its Find*.cmake file
```
Comment (coryan) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:3105]: You can avoid this by using crc32c 1.1.1:

https://github.com/google/crc32c/commit/b377ce46f56823cf38dd7045f4997dd4b5c6d04c

File: ci/conda_env_unix.txt (hunk 4 above)
Comment (coryan) [reply ci/conda_env_unix.txt:23]: I am one of the authors of google's GCS SDK.  Yes, the SDK definitely depends on `nlohmann_json`.

Comment (pitrou) [reply ci/conda_env_unix.txt:23]: Conda has dependency handling, so there is no reason to mention it explicitly, though.

**Comment by josiahyan** [comment]
I’ve discussed this PR with @coryan , and we think its better to start over, given the need to upgrade dependencies, and handle the Abseil libraries (which I believe, when used with gRPC are causing the build to fail on some platforms) with more scripting. He will be taking over this work.

**Pull request closed without merging**